### ISSUE TITLE
[feature request] Stochastic Variance Reduced Gradient (SVRG) optimizer

### ISSUE THREAD

**Issue by AnesBenmerzoug** [pytorch/pytorch#6657]
I'm currently working on neural networks for my Master's thesis and I stumbled upon the optimizer described in this paper:
https://papers.nips.cc/paper/4937-accelerating-stochastic-gradient-descent-using-predictive-variance-reduction.pdf
and I managed to make an implementation for myself and I thought I could contribute it if it is accepted by the community.
There is one thing about this optimizer that makes it a bit tougher than others. And that is the fact that it uses two sets of parameters ( which means two models in my implementation ): the main ones and a snapshot of the main taken every few epochs, and that are used to reduce the variance of the optimization.

**Comment by Breakend** [comment]
This would be a nice to have actually if you have time to get to it! +1

**Comment by AnesBenmerzoug** [comment]
Sure, I will start working on it. The optimizer by itself is ready, I just have to integrate it with the rest of the code, test it more thoroughly and then make a pull request.
